pkg/operator/credentials.go

```diff
@@ -0,0 +1,159 @@
+/*
+Copyright 2023.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package operator
+
+// NOTE: this file is influenced by the new approach of how to do STS described in
+// https://docs.google.com/document/d/1iFNpyycby_rOY1wUew-yl3uPWlE00krTgr9XHDZOTNo.
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws/arn"
+	cco "github.com/openshift/cloud-credential-operator/pkg/apis/cloudcredential/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	operatorCredentialsRequestName = "aws-load-balancer-operator"
+	operatorCredentialsSecretName  = "aws-load-balancer-operator"
+	operatorServiceAccountName     = "aws-load-balancer-operator-controller-manager"
+	roleARNEnvVar                  = "ROLEARN"
+	webIdentityTokenPath           = "/var/run/secrets/openshift/serviceaccount/token"
+	credentialsKey                 = "credentials"
+	waitForSecretTimeout           = 5 * time.Minute
+	waitForSecretPollInterval      = 5 * time.Second
+)
+
+// ProvisionCredentials provisions cloud credentials secret in the given namespace
+// with IAM policies required by the operator. The credentials data are put
+// into a file which can be used to setup AWS SDK client.
+func ProvisionCredentials(ctx context.Context, client client.Client, secretNamespace string) (string, error) {
+	roleARN := os.Getenv(roleARNEnvVar)
+	if roleARN != "" && !arn.IsARN(roleARN) {
+		return "", fmt.Errorf("provided role arn is invalid: %q", roleARN)
+	}
+
+	credReq := buildCredentialsRequest(secretNamespace, operatorCredentialsSecretName, roleARN)
+
+	// create credentials request resource
+	if err := client.Create(ctx, credReq); err != nil {
```

Comment:
I ran into an issue. When testing this PR with AWS STS, I forgot to put the the `ROLEARN` in the manager deployment yaml as a ENV variable. I recreated the deployment with `ROLEARN`, but scratched my head why it still wasn't working.

Then I realized it doesn't compare the content of the credential request, but just ensures an object exists with the same name. So I had to go delete the CR manually to get it to work.

Not saying it needs to compare, but do you think users would benefit from a comparison? Just curious if you considered these scenarios and if you are okay with the implications.

Reply:
> Not saying it needs to compare, but do you think users would benefit from a comparison? Just curious if you considered these scenarios and if you are okay with the implications.

I thought of this scenario. In a slightly different context but with the same result - what if the role ARN needs to be updated? I implemented it the way CCO people proposed in their [How to STS document](https://docs.google.com/document/d/1iFNpyycby_rOY1wUew-yl3uPWlE00krTgr9XHDZOTNo) because it's simple and agreed by the CCO people and reviewers of the document. However I'm still not 100% convinced because it's easy to get it wrong and end up in a situation when you have to do manual repairs.
Another point is the complexity of the comparison approach. Apart from the comparison itself which is a little more complex but still fine (as we have a lot of similar examples), there is the aspect of the secret's contents. If `CredentialsRequest` is updated, ALBO needs to make sure it reads the new contents not the old ones from the previous `CredentialsRequest` version.

Reply:
Slack discussion: https://redhat-internal.slack.com/archives/C04KRSVAZ37/p1695799843775839.

Reply:
Yes, they seem to like the approach and will update the reference doc accordingly. Let me change the part then to add a comparison+update of credentialsrequest.

Reply:
Done in https://github.com/openshift/aws-load-balancer-operator/pull/109/commits/f1a89d8c1a3bd2b1d39eaad8d9f06e8b6866ca1a.